dsp/src/lockin.rs

```diff
@@ -0,0 +1,544 @@
+                    - self.timestamps[0].unwrap())
+                    * self.harmonic as i32;
+                let phase = self.phase_offset
+                    + 2. * PI * integer_phase as f32 / reference_period as f32;
+                let (sine, cosine) = libm::sincosf(phase);
+                let sample = *sample as f32;
+                *i = sine * sample;
+                *q = cosine * sample;
+            });
+
+        Ok((in_phase, quadrature))
+    }
+
+    /// Filter the in-phase and quadrature signals using the supplied
+    /// biquad IIR. The signal arrays are modified in place.
+    ///
+    /// # Arguments
+    ///
+    /// * `in_phase` - In-phase signal.
+    /// * `quadrature` - Quadrature signal.
+    pub fn filter(&mut self, in_phase: &mut [f32], quadrature: &mut [f32]) {
+        in_phase
+            .iter_mut()
+            .zip(quadrature.iter_mut())
+            .for_each(|(i, q)| {
+                *i = self.iir[0].update(&mut self.iirstate[0], *i);
+                *q = self.iir[1].update(&mut self.iirstate[1], *q);
+            });
+    }
+}
+
+/// Decimate the in-phase and quadrature signals to
+/// `DECIMATED_BUFFER_SIZE`. The ratio of `ADC_SAMPLE_BUFFER_SIZE` to
+/// `DECIMATED_BUFFER_SIZE` must be a power of 2.
+///
+/// # Arguments
+///
+/// * `in_phase` - In-phase signal.
+/// * `quadrature` - Quadrature signal.
+///
+/// # Returns
+///
+/// The decimated in-phase and quadrature signals.
+pub fn decimate(
+    in_phase: [f32; ADC_SAMPLE_BUFFER_SIZE],
+    quadrature: [f32; ADC_SAMPLE_BUFFER_SIZE],
+) -> ([f32; DECIMATED_BUFFER_SIZE], [f32; DECIMATED_BUFFER_SIZE]) {
+    let n_k = ADC_SAMPLE_BUFFER_SIZE / DECIMATED_BUFFER_SIZE;
+    debug_assert!(
+        ADC_SAMPLE_BUFFER_SIZE == DECIMATED_BUFFER_SIZE || n_k % 2 == 0
+    );
+
+    let mut in_phase_decimated = [0f32; DECIMATED_BUFFER_SIZE];
+    let mut quadrature_decimated = [0f32; DECIMATED_BUFFER_SIZE];
+
+    in_phase_decimated
+        .iter_mut()
+        .zip(quadrature_decimated.iter_mut())
+        .zip(
+            in_phase
+                .iter()
+                .step_by(n_k)
+                .zip(quadrature.iter().step_by(n_k)),
+        )
+        .for_each(|((i_d, q_d), (i, q))| {
+            *i_d = *i;
+            *q_d = *q;
+        });
+
+    (in_phase_decimated, quadrature_decimated)
+}
+
+/// Compute the magnitude and phase from the in-phase and quadrature
+/// signals. The in-phase and quadrature arrays are modified in place.
+///
+/// # Arguments
+///
+/// * `in_phase` - In-phase signal.
+/// * `quadrature` - Quadrature signal.
+pub fn magnitude_phase(in_phase: &mut [f32], quadrature: &mut [f32]) {
+    in_phase
+        .iter_mut()
+        .zip(quadrature.iter_mut())
+        .for_each(|(i, q)| {
+            let new_i = libm::sqrtf([*i, *q].iter().map(|i| i * i).sum());
+            let new_q = libm::atan2f(*q, *i);
+            *i = new_i;
+            *q = new_q;
+        });
+}
```

Comment:
Is the goal here to hold onto/pass around the iterator? I don't believe it's possible to `collect()` into an array.

Reply:
As mentioned above, passing iterators around might be tricky because the types become a bit complicated.
Let's just provide the building blocks for single-sample processing and then have the user use iterators in `process()` (the ISR in `main.rs`) where and when convenient. Like https://github.com/quartiq/stabilizer/pull/177#pullrequestreview-538698163